src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -0,0 +1,1001 @@
+//     overhead effort:
+//       i) For each spanned cluster, we have to look up the start of
+//          the crossing object.
+//      ii) Each time we scan the very large object, we have to
+//          sequentially walk through its pointer location
+//          descriptors, skipping over all of the pointers that
+//          precede the start of the range of addresses that we
+//          consider relevant.
+
+
+// Because old-gen heap memory is not necessarily contiguous, and
+// because cards are not necessarily maintained for young-gen memory,
+// consecutive card numbers do not necessarily correspond to consecutive
+// address ranges.  For the traditional direct-card-marking
+// implementation of this interface, consecutive card numbers are
+// likely to correspond to contiguous regions of memory, but this
+// should not be assumed.  Instead, rely only upon the following:
+//
+//  1. All card numbers for cards pertaining to the same
+//     ShenandoahHeapRegion are consecutively numbered.
+//  2. In the case that neighboring ShenandoahHeapRegions both
+//     represent old-gen memory, the card regions that span the
+//     boundary between these neighboring heap regions will be
+//     consecutively numbered.
+//  3. (A corollary) In the case that an old-gen object straddles the
+//     boundary between two heap regions, the card regions that
+//     correspond to the span of this object will be consecutively
+//     numbered.
+//
+// ShenandoahCardCluster abstracts access to the remembered set
+// and also keeps track of crossing map information to allow efficient
+// resolution of object start addresses.
+//
+// ShenandoahCardCluster supports all of the services of
+// DirectCardMarkRememberedSet, plus it supports register_object() and lookup_object().
+// Note that we only need to register the start addresses of the object that
+// overlays the first address of a card; we need to do this for every card.
+// In other words, register_object() checks if the object crosses a card boundary,
+// and updates the offset value for each card that the object crosses into.
+// For objects that don't straddle cards, nothing needs to be done.
+//
+class ShenandoahCardCluster: public CHeapObj<mtGC> {
+
+private:
+  ShenandoahDirectCardMarkRememberedSet* _rs;
+
+public:
+  static const size_t CardsPerCluster = 64;
+
+private:
+  typedef struct cross_map { uint8_t first; uint8_t last; } xmap;
+  typedef union crossing_info { uint16_t short_word; xmap offsets; } crossing_info;
+
+  // ObjectStartsInCardRegion bit is set within a crossing_info.offsets.start iff at least one object starts within
+  // a particular card region.  We pack this bit into start byte under assumption that start byte is accessed less
+  // frequently than last byte.  This is true when number of clean cards is greater than number of dirty cards.
+  static const uint16_t ObjectStartsInCardRegion = 0x80;
+  static const uint16_t FirstStartBits           = 0x7f;
```

Comment:
I see these are used to do operations against `uint8_t` (`....offsets.first`). Any reason why these should not be `uint8_t`?